aws/volume.go

```diff
@@ -0,0 +1,196 @@
+package aws
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/ec2"
+	"github.com/giantswarm/backoff"
+	"github.com/giantswarm/microerror"
+)
+
+type EBSConfig struct {
+	AWSInstanceID string
+	AwsSession    *session.Session
+	DeviceName    string
+	ForceDetach   bool
+	TagKey        string
+	TagValue      string
+}
+
+type EBS struct {
+	awsInstanceID string
+	awsSession    *session.Session
+	deviceName    string
+	forceDetach   bool
+	tagKey        string
+	tagValue      string
+}
+
+func NewEBS(config EBSConfig) (*EBS, error) {
+	if config.AWSInstanceID == "" {
+		return nil, microerror.Maskf(invalidConfigError, "config.AWSInstanceID must not be empty")
+	}
+	if config.AwsSession == nil {
+		return nil, microerror.Maskf(invalidConfigError, "config.AwsSession must not be nil")
+	}
+	if config.DeviceName == "" {
+		return nil, microerror.Maskf(invalidConfigError, "config.DeviceName must not be empty")
+	}
+	if config.TagKey == "" {
+		return nil, microerror.Maskf(invalidConfigError, "config.TagKey must not be empty")
+	}
+	if config.TagValue == "" {
+		return nil, microerror.Maskf(invalidConfigError, "config.TagValue must not be empty")
+	}
+
+	newEBS := &EBS{
+		awsInstanceID: config.AWSInstanceID,
+		awsSession:    config.AwsSession,
+		deviceName:    config.DeviceName,
+		forceDetach:   config.ForceDetach,
+		tagKey:        config.TagKey,
+		tagValue:      config.TagValue,
+	}
+	return newEBS, nil
+}
+
+func (s *EBS) AttachByTag() error {
+	// create ec2 client
+	ec2Client := ec2.New(s.awsSession)
+
+	volume, err := s.describe(ec2Client)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+	fmt.Printf("Fetched volume-id '%s'\n", *volume.VolumeId)
+
+	if *volume.State == ec2.VolumeStateInUse &&
+		len(volume.Attachments) == 1 &&
+		*volume.Attachments[0].InstanceId == s.awsInstanceID {
+		fmt.Printf("Volume is already attached to this instance. Nothing to do.\n")
+		return nil
+	} else if *volume.State == ec2.VolumeStateInUse {
+		fmt.Printf("Volume is attached to %q and is in state %q. Trying detach the volume\n", *volume.Attachments[0].InstanceId, *volume.State)
+
+		err := s.detach(ec2Client, volume)
+		if err != nil {
+			return microerror.Mask(err)
+		}
+	} else {
+		fmt.Printf("Volume state is %q.\n", *volume.State)
```

Comment:
I'm wondering if the code shouldn't error in this case.

Reply:
this means that volume is `not-used(avaiable)` state so it can be attached without detaching. this is very possible scenario and should not be a problem,